### ISSUE TITLE
Restrict the maximum number of parallel tasks in ParallelExecutionStrategy

### ISSUE THREAD

**Issue by teo78** [graphql-dotnet/graphql-dotnet#1239]
## Summary

We're using the GraphQL dotnet library (version 2.4) in a ASP .NET Core application using MongoDB as the database. We are experiencing extreme performance drops occasionally when running large queries involving a lot of database queries. The issue was found to be due to the MongoDB driver exhausting the connection pool (Default = 100), in which case db queries started to queue up and run extremely slow. As a band-aid solution increasing the connection pool on the MongoDB driver solved the issue.

However, this raises the question of whether the library should have a limit to how many parallel nodes it will execute simultaneously.

### Test using PR: 'Optimize ParallelExecutionStrategy: Mark2'
I did test the same problematic queries using the modifications in the 'Optimize ParallelExecutionStrategy: Mark2' pull request #976. Execution was often even slower, which is not unexpected since it runs even more nodes in parallel.

By modifying the code in this PR to include a maximum number of parallel tasks, the query runs fast. I.e. with the default max connection pool of 100, we do not see the performance drops and do not have to increase the max pool size to alleviate the problem.

The fix tested was by changing the 'pendingNodes' from a List to a Queue and limiting the number of current tasks instead of adding all unrestricted:

```csharp
// Start executing all pending nodes, limiting the max number of current tasks
while (currentTasks.Count() < _taskLimit && pendingNodes.TryDequeue(out var pendingNode))
{
    var task = ExecuteNodeAsync(context, pendingNode);
    currentTasks.AddLast(task);
}
```

### Motivation

Limit the number of maximum parallel nodes executed, keeping it unrestricted can cause issues be starting to many parallel database calls, threads, etc.

**Comment by joemcbride** [comment]
@teo78 Thanks for adding this issue!  Would you mind providing an initial PR with this support?

**Comment by teo78** [comment]
I'll have a go at it. Regarding setting of the maximum parallel limit, my thinking is this:

1. Add a 'MaxParallelExecutionLimit' (Or Similar) int property to the 'ExecutionOptions' class.
    - Default value: int.MaxValue (to keep it unrestricted).
2. In 'ExecuteAsync' method of the 'DocumentExecuter' class, pass this setting to the 'BuildeExecutionContext'
   - This means also adding 'MaxParallelExecutionLimit' property to the 'ExectionContext' class
3. 'ParallelExecutionStrategy' now has access to the setting via the 'context' parameter of the 'ExecuteNodeTreeAsync' method, and logic to enforce the limit can be implemented there.

Should I base the PR on the Master branch or the branch from the other PR to improve the ParallelExecutionStrategy (#976) ?

**Comment by sungam3r** [comment]
It seems to me that it is better to put such changes in a separate PR.

> Default value: int.MaxValue (to keep it unrestricted).

Or just nullable int and null as unrestricted.

**Comment by teo78** [comment]
Took a stab at this and created a PR (#1251)

**Comment by sungam3r** [comment]
Closed as resolved in #1251
